Fix ARM build process and environment setup

jakule opened the pull request:
Fix the ARM build process by using the Debian 11-based Docker image instead of the CI image that requires a newer GLibc version.

Build https://drone.platform.teleport.sh/gravitational/teleport/27432

jakule committed:
Refactor ARM build process and environment setup

This commit changes the ARM build process by updating the build tools. The `CC` environment variable in Makefile was changed from `arm-linux-gnueabihf-gcc` to `arm-linux-gnueabi-gcc` to match the new setup.
Our Dockerfile-arm has been rewritten to use Debian 11 image instead of re-using CI Ubuntu image to match glibc version on distroless images.

jakule committed:
Fix ARM build process and environment setup

Fixed the ARM build process by updating the C compiler set in the Makefile and simplifying Dockerfile dependencies for the build. De-cluttered Makefile for streamlining build process. Replaced the usage of an incorrect C compiler in the Makefile and reduced the number of dependencies in Dockerfile-arm to only those necessary for building on ARM. Fixed build errors that prevented successful compilation on ARM.

jakule committed:
Remove RUST_VERSION as it's not being used in the ARM Dockerfile

jakule commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,72 @@
-ARG BUILDBOX_VERSION
-FROM public.ecr.aws/gravitational/teleport-buildbox:$BUILDBOX_VERSION
+FROM docker.io/library/debian:11
```
> Do we have any documentation around our build process and which artifacts are built from which docker image?

We are working on it. We should have something later this month.

> In what circumstances do we use debian instead of the buildbox and why?

"buildbox" as a name is confusing. We currently have multiple buildboxes. The buildbox that you're referring to uses Ubuntu 22.04 image and should be only used in our CI and not for any releases. We're using Debian 11 here as our binaries must run on distroless which is based on Debian 11.

> Why is the buildbox not just using debian?

Because our CI buildbox should be ONLY used in CI and nowhere else. I want to standardize our Docker images, but in order to do that we need to move out, all builds out of the builbox. These images should not depend on each order, as changes to CI should not break the release (which happened recently).

codingllama replied:
Capturing some of what you said, like the purpose of the image, in a comment here would be nice.

ibeckermayer commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,72 @@
+    LC_CTYPE="en_US.UTF-8" \
+    DEBIAN_FRONTEND="noninteractive"
+
+# BUILDARCH is automatically set by DOCKER when building the image with Build Kit (MacOS by deafult).
```
```suggestion
# BUILDARCH is automatically set by DOCKER when building the image with Build Kit (MacOS by default).
```

jakule committed:
Fix typo

Co-authored-by: Isaiah Becker-Mayer <isaiah@goteleport.com>

codingllama commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,72 @@
+        unzip \
+        zip \
+        # ARM dependencies
+        libc6-dev-armhf-cross \
+        gcc-arm-linux-gnueabihf \
```
```suggestion
        gcc-arm-linux-gnueabihf \
        libc6-dev-armhf-cross \
```

jakule replied:
Changed

codingllama commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,72 @@
+ARG NODE_VERSION
+ENV NODE_PATH="/usr/local/lib/nodejs-linux"
+ENV PATH="$PATH:${NODE_PATH}/bin"
+RUN export NODE_ARCH=$(if [ "$BUILDARCH" = "amd64" ]; then echo "x64"; else echo "arm64"; fi) && \
+    export NODE_URL="https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-${NODE_ARCH}.tar.xz" && \
+    mkdir -p ${NODE_PATH} && \
+    curl -o /tmp/nodejs.tar.xz -fsSL ${NODE_URL} && \
+    tar -xJf /tmp/nodejs.tar.xz -C /usr/local/lib/nodejs-linux --strip-components=1
```
nit: Bash style, proper quoting, exports unnecessary.

```suggestion
RUN NODE_ARCH="$(if [ '$BUILDARCH' = 'amd64' ]; then echo 'x64'; else echo 'arm64'; fi)" && \
    NODE_URL="https://nodejs.org/dist/v$NODE_VERSION/node-v$NODE_VERSION-linux-$NODE_ARCH.tar.xz" && \
    mkdir -p "$NODE_PATH" && \
    curl -o /tmp/nodejs.tar.xz -fsSL "$NODE_URL" && \
    tar -xJf /tmp/nodejs.tar.xz -C /usr/local/lib/nodejs-linux --strip-components=1
```

jakule replied:
"proper quoting" around `"$BUILDARCH"` breaks it as you want to evaluate the value and not take the literal. Applied everything else.

codingllama commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,72 @@
+RUN export NODE_ARCH=$(if [ "$BUILDARCH" = "amd64" ]; then echo "x64"; else echo "arm64"; fi) && \
+    export NODE_URL="https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-${NODE_ARCH}.tar.xz" && \
+    mkdir -p ${NODE_PATH} && \
+    curl -o /tmp/nodejs.tar.xz -fsSL ${NODE_URL} && \
```
> /tmp/nodejs.tar.xz

Use mktemp, or pipe curl into tar like below?

jakule replied:
Done

codingllama commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,72 @@
+
+# Install Go.
+ARG GOLANG_VERSION
+RUN mkdir -p /opt && cd /opt && curl -fsSL https://storage.googleapis.com/golang/$GOLANG_VERSION.linux-${BUILDARCH}.tar.gz | tar xz && \
+    mkdir -p /go/src/github.com/gravitational/teleport && \
+    chmod a+w /go && \
+    chmod a+w /var/lib && \
+    chmod a-w /
```
Quoting, 1 command per line.

```suggestion
RUN mkdir -p /opt && \
    cd /opt && \
    curl -fsSL "https://storage.googleapis.com/golang/$GOLANG_VERSION.linux-$BUILDARCH.tar.gz" | tar xz && \
    mkdir -p /go/src/github.com/gravitational/teleport && \
    chmod a+w /go && \
    chmod a+w /var/lib && \
    chmod a-w /
```

jakule replied:
Changed

codingllama commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,72 @@
+    GOROOT="/opt/go" \
+    PATH="$PATH:/opt/go/bin:/go/bin:/go/src/github.com/gravitational/teleport/build"
+
+# Add the CI user.
```
User switching  😢

codingllama commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,72 @@
+# Add the CI user.
+ARG UID
+ARG GID
+RUN groupadd ci --gid=$GID -o && \
```
If this is not meant for CI, why is the "ci" user here?

jakule replied:
The name doesn't matter. The user must have UID=1000 https://github.com/gravitational/teleport/blob/5d82fe6b1f5d80266d2de4b28a9b0f4e5519d1c3/.drone.yml#L4090-L4101
I can rename it, but all our scripts expect user `ci` with `UID=1000`. Changing it now may break something. I think we should re-visit the topic after we move out of Drone.

codingllama replied:
I'd take a comment here too, as this strongly suggests the image is used for CI.

jakule committed:
Address code review comments

reedloden commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,75 @@
-FROM public.ecr.aws/gravitational/teleport-buildbox:$BUILDBOX_VERSION
+# This Dockerfile is used to build Teleport on ARM only.
+# We are using the official Debian 11 image as a base image
+# because the finaly binary must be compatible with distroless
```
```suggestion
# because the final binary must be compatible with distroless
```

jakule replied:
Thx

reedloden commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,75 @@
+# We are using the official Debian 11 image as a base image
+# because the finaly binary must be compatible with distroless
+# images that are also Debian 11 based: https://github.com/GoogleContainerTools/distroless
+FROM docker.io/library/debian:11
```
We should make sure the GHA workflows are using `docker/login-action` to authenticate to Docker Hub so we don't run into any rate limiting issues.

jakule replied:
This docker file is used by Drone (at least for now), so I'm not sure if this applies.

camscale replied:
It will be used on github too until everything is migrated to a common centos7 buildbox (or whatever replaces that), but the workflow in which I will use this has the dockerhub login so should be fine.

jakule committed:
Fix typo

Co-authored-by: Reed Loden <reed@goteleport.com>

codingllama commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,75 @@
+    NODE_FILE="$(mktemp node-XXXXXX.tar.xz)" && \
+    mkdir -p "$NODE_PATH" && \
+    curl -o "$NODE_FILE" -fsSL "$NODE_URL" && \
+    tar -xJf "$NODE_FILE" -C /usr/local/lib/nodejs-linux --strip-components=1
```
```suggestion
    tar -xJf "$NODE_FILE" -C /usr/local/lib/nodejs-linux --strip-components=1 && \
    rm -f "$NODE_FILE"
```

codingllama commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,75 @@
+ARG NODE_VERSION
+ENV NODE_PATH="/usr/local/lib/nodejs-linux"
+ENV PATH="$PATH:${NODE_PATH}/bin"
+RUN NODE_ARCH=$(if [ "$BUILDARCH" = 'amd64' ]; then echo 'x64'; else echo 'arm64'; fi) && \
```
I don't see why this would cause problems:

```suggestion
RUN NODE_ARCH="$(if [ "$BUILDARCH" = 'amd64' ]; then echo 'x64'; else echo 'arm64'; fi)" && \
```

Seems to work fine in my clone of your branch.

jakule replied:
I missed that. The outer quotes are working, only the inner ones are problematic :sweat_smile:

codingllama commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,75 @@
+ARG GOLANG_VERSION
+RUN mkdir -p /opt && \
+    cd /opt && \
+    curl -fsSL https://storage.googleapis.com/golang/$GOLANG_VERSION.linux-${BUILDARCH}.tar.gz | tar xz && \
```
```suggestion
    curl -fsSL "https://storage.googleapis.com/golang/$GOLANG_VERSION.linux-$BUILDARCH.tar.gz" | tar xz && \
```

codingllama commented on build.assets/Dockerfile-arm:
```diff
@@ -1,8 +1,75 @@
+# Add the CI user.
+ARG UID
+ARG GID
+RUN groupadd ci --gid=$GID -o && \
+    useradd ci --uid=$UID --gid=$GID --create-home --shell=/bin/sh && \
```
```suggestion
RUN groupadd ci --gid="$GID" -o && \
    useradd ci --uid="$UID" --gid="$GID" --create-home --shell=/bin/sh && \
```

codingllama reviewed (approved):
Still missing some quoting, but otherwise alright.

jakule committed:
Update Dockerfile to clean up temporary node file and improve variable usage

The Dockerfile has been updated to delete the temporary node file after use to reduce unnecessary space usage. Additionally, run commands have been adjusted to use double quotes around variables for better compatibility and safety. Lastly, a comment is added to explain the reason behind the persistence of CI user creation in the image.
